### DISCUSSION TITLE
Are commands executed multiple times?

### DISCUSSION THREAD

**Discussion by callegar** [lvjr/tabularray#65] [Q&A]
Hi,

the `cmd=` parameter to have a command executed on cells content, initially appears as a nice candidate for accumulating values (e.g., to compute row totals). Unfortunately, it seems to be executed multiple times, so that the accumulators accumulate more than expected.

- Is this actually the case?
- Is there any way to assure that a command is executed only once?

Thanks for any suggestion

**Comment by lvjr** [comment]
Yes, just like `tabularx` package, `tabularray` need to measure cells first before finally building the whole table, hence `cmd`s will be executed multiple times.

If you need to do cell calculations, I recommend to use `spreadtab` package together with `tabularray`.

**Comment by callegar** [comment]
`spreadtab` looks indeed to be a solution, thanks. May I still suggest that `cmd` is paired with some `cmd-once` in future versions of the package?

**Reply by davidcarlisle** [reply]
Usually (often) it's not enough to only execute the command once (eg it may need to produce some approximation of the final output on trial runs so that the width measurement does something sensible).

What would be good though (and I wish I had added to `tabularx`) is a boolean so that client code can test if it is on the final typesetting run. AMS alignments have `\ifmeasuring@`  but in `tabularx` I don't offer a `\newif` token so you need tests like `\ifx\hbadness\@tempcnta trial run \else final run \fi` which works but is a pretty poor interface.

**Reply by callegar** [reply]
That's right, however if you want to do non-idempotent operations, executing only once is necessary.  These operations typically need to be done on the *raw* cell value, so doing them at the first run would probably be enough for most tasks.

A useful alternative would be to make an invocation counter available to the command, so that it can be tested in the command.
Incidentally, it would be great to also have the row and column indexes available in the command.

**Reply by lvjr** [reply]
In `tabularray`, `cmd` option is designed for changing current cell text. If the cell text is different in different typesetting runs, you will probably get wrong table layout.

For complicated table contents, I would suggest to build the contents and write them into a macro before `tblr` environment and then use `expand` option to expand the macro in the `tblr` environment.

**Reply by lvjr** [reply]
In `tabularray`, two counters `rownum` and `colnum` are always available in typesetting cells and can be used by `cmd` option.

**Reply by callegar** [reply]
I agree with you, however, many of the proposed solutions are really a bit invasive for tiny things.  For instance, `spreadtab` has the distinction between text and number cells, which implies the introduction of the text marker `@` in many many cells. Building the content in advance is also very general, however to merely build an accumulator it is a bit cumbersome.

If the invocation index of the command was made available (e.g., as `\invocation_idx`) one could test it in the command, and then it would be the user responsibility not to do crazy things with it.  Obviously with the rownum and colnum it should be easy to create such index.

Alternatively, may I modify my suggestion to a `cmd-once` that must return *nothing*, so that the content of the cell is not altered from it?

**Reply by lvjr** [reply]
You can define run-once commands like this:
````latex
\gdef\mycmd#1{%
  do something and store the result
  \gdef\mycmd#1{get stored result}%
}
````

**Reply by cdacl** [reply]
This method will make the cell width not to be measured correctly.

```
\gdef\mycmd{do something and store the result\gdef\mycmd{get stored result}}
\begin{tblr}{
 hlines,
 vlines,
}
\mycmd Alpha   & Beta   & Gamma  \\
 Epsilon & Zeta   & Eta    \\ 
 Iota    & Kappa  & Lambda \\
\end{tblr} 
```
The output:

![1](https://user-images.githubusercontent.com/87848848/128977138-a078e61d-0513-42a2-9fc3-67cd430d7b74.png)

The width of first cell was set to the width of 'do something and store the resultAlpha'.

**Reply by callegar** [reply]
My idea was actually have commands executed once *per cell*. Currently, because I have a single column where I need this, I am checking on a row basis with something like this:

```LaTeX
\fp_new:N \g_x_coins_fp
\NewDocumentCommand { \CountCoins } { m }
  {
    \tl_if_blank:nF { #1 }
      {
        \tl_if_exist:cF { g_x_done_ \therownum _tl }
        { 
          \fp_gadd:Nn \g_x_coins_fp { #1 }
          \tl_gset:cn { g_x_done_ \therownum _tl } { \relax }
        }
      }
    #1
  }
```
which seems to work well, but dirties the global command space with a lot of token lists.  The rationale supporting my suggestion of a cmd-once was to simplify this kind of tasks.
